### PR TITLE
Light client proof bitcoin e2e test

### LINKED ISSUE #1466: Light client proof e2e tests

### PR DESCRIPTION
**Pull request by ercecan**
# Description
Tests light client proof with bitcoin e2e framework

## Linked Issues
- Fixes #1466

### CODE REVIEW

File: bin/citrea/tests/bitcoin_e2e/light_client_test.rs [hunk 1]
```diff
@@ -143,3 +149,317 @@ async fn test_light_client_proving() -> Result<()> {
...
+                .sum();
+            let borshed_state_diff = borsh::to_vec(state_diff).unwrap();
+            let compressed_state_diff =
+                citrea_primitives::compression::compress_blob(&borshed_state_diff);
+            println!(
+                "StateDiff: size {}, compressed {}",
+                state_diff_size,
+                compressed_state_diff.len()
+            );
+        }
```
Comment (eyusufatik) [inline bin/citrea/tests/bitcoin_e2e/light_client_test.rs]: why?

File: bin/citrea/tests/bitcoin_e2e/light_client_test.rs [hunk 2]
```diff
@@ -143,3 +149,317 @@ async fn test_light_client_proving() -> Result<()> {
...
+
+        // Assert that commitments are queryable
+        let commitments = batch_prover
+            .client
+            .http_client()
+            .get_sequencer_commitments_on_slot_by_number(commitment_l1_height)
+            .await
+            .unwrap()
+            .unwrap();
+        assert_eq!(commitments.len(), 4);
```
Comment (eyusufatik) [inline bin/citrea/tests/bitcoin_e2e/light_client_test.rs]: can you fix this thing on other tests as well. remmeber our realization about commit / reveal

File: bin/citrea/tests/bitcoin_e2e/light_client_test.rs [hunk 3]
```diff
@@ -143,3 +149,317 @@ async fn test_light_client_proving() -> Result<()> {
...
+        assert_ne!(
+            light_client_proof3.light_client_proof_output.last_l2_height,
+            light_client_proof.light_client_proof_output.last_l2_height
+        );
+
+        assert_ne!(
+            light_client_proof3.light_client_proof_output.state_root,
+            light_client_proof.light_client_proof_output.state_root
+        );
+
```
Comment (eyusufatik) [inline bin/citrea/tests/bitcoin_e2e/light_client_test.rs:445]: can you check if any non-stitchable batch proof tuples were put into the output vec?

**Review by eyusufatik — CHANGES_REQUESTED** [review]

File: bin/citrea/tests/bitcoin_e2e/light_client_test.rs [hunk 4]
```diff
@@ -143,3 +149,315 @@ async fn test_light_client_proving() -> Result<()> {
...
+        let rpc_bind_host = sequencer.config().rpc_bind_host();
+        let rpc_bind_port = sequencer.config().rpc_bind_port();
+
+        let socket_addr = SocketAddr::new(rpc_bind_host.parse().unwrap(), rpc_bind_port);
+
+        let seq_test_client = make_test_client(socket_addr).await.unwrap();
+
+        // publish min_soft_confirmations_per_commitment confirmations
+        for _ in 0..4 * min_soft_confirmations_per_commitment {
```
Comment (jfldde) [inline bin/citrea/tests/bitcoin_e2e/light_client_test.rs]:
```suggestion
        for _ in 0..min_soft_confirmations_per_commitment * n_conf {
```
Can lift 4 to a var and re-use on the * 4 and 8 below

File: bin/citrea/tests/bitcoin_e2e/light_client_test.rs [hunk 5]
```diff
@@ -143,3 +149,315 @@ async fn test_light_client_proving() -> Result<()> {
...
+            batch_proofs[1].proof_output.final_state_root
+        );
+
+        assert!(light_client_proof
+            .light_client_proof_output
+            .unchained_batch_proofs_info
+            .is_empty());
+
+        // Generate another da block so we generate another lcp
+        da.generate(1u64, None).await?;
```
Comment (jfldde) [inline bin/citrea/tests/bitcoin_e2e/light_client_test.rs]:
```suggestion
        da.generate(1, None).await?;
```

File: bin/citrea/tests/bitcoin_e2e/sequencer_commitments.rs [hunk 6]
```diff
@@ -276,7 +276,7 @@ impl TestCase for SequencerSendCommitmentsToDaTest {
             .await?;
 
         // Wait for blob tx to hit the mempool
-        da.wait_mempool_len(1, None).await?;
```
Comment (jfldde) [inline bin/citrea/tests/bitcoin_e2e/sequencer_commitments.rs:279]: Good fix, might explain some flakiness

File: bin/citrea/tests/bitcoin_e2e/light_client_test.rs [hunk 7]
```diff
@@ -206,8 +206,10 @@ impl TestCase for LightClientProvingTestMultipleProofs {
 
         let seq_test_client = make_test_client(socket_addr).await.unwrap();
 
+        let mut n_commitments = 4;
```
Comment (jfldde) [inline bin/citrea/tests/bitcoin_e2e/light_client_test.rs]: Doesn't need to be mut

**Review by jfldde — APPROVED** [review]

**Review by eyusufatik — APPROVED** [review]

File: bin/citrea/tests/bitcoin_e2e/light_client_test.rs [hunk 8]
```diff
@@ -143,3 +150,325 @@ async fn test_light_client_proving() -> Result<()> {
...
+            with_batch_prover: true,
+            with_light_client_prover: true,
+            with_full_node: true,
+            ..Default::default()
+        }
+    }
+
+    fn sequencer_config() -> SequencerConfig {
+        SequencerConfig {
+            min_soft_confirmations_per_commitment: 20,
```
Comment (jfldde) [inline bin/citrea/tests/bitcoin_e2e/light_client_test.rs:167]: What's the reason for using 20 here ?

File: bin/citrea/tests/bitcoin_e2e/light_client_test.rs [hunk 9]
```diff
@@ -143,3 +150,325 @@ async fn test_light_client_proving() -> Result<()> {
...
+            da_update_interval_ms: 500,
+            mempool_conf: SequencerMempoolConfig {
+                pending_tx_limit: 100_000,
+                pending_tx_size: 2000,
+                queue_tx_limit: 100_000,
+                queue_tx_size: 200,
+                base_fee_tx_limit: 100_000,
+                base_fee_tx_size: 200,
+                max_account_slots: 1600,
+            },
```
Comment (jfldde) [inline bin/citrea/tests/bitcoin_e2e/light_client_test.rs]:
```suggestion
            mempool_conf: SequencerMempoolConfig {
                pending_tx_size: 2000,
                max_account_slots: 1600,
                ..Default::default()
            },
```

File: bin/citrea/tests/bitcoin_e2e/light_client_test.rs (hunk 8 above)
Comment (ercecan) [reply bin/citrea/tests/bitcoin_e2e/light_client_test.rs:167]: trying to land multiple proofs in a da block, so I need the commitments to be big enough to create two batch proofs

File: crates/primitives/src/constants.rs [hunk 10]
```diff
@@ -21,7 +21,7 @@ const fn get_reveal_light_client_prefix() -> &'static [u8] {
 
 const fn get_max_txbody_size() -> usize {
     match option_env!("CI_TEST_MODE") {
-        Some(v) if matches!(v.as_bytes(), b"1" | b"true") => 39700,
```
Comment (yaziciahmet) [inline crates/primitives/src/constants.rs]: test_sequnecer_commitment_threshold test's tx counts should also be updated accordingly.

File: bin/citrea/tests/bitcoin_e2e/light_client_test.rs [hunk 11]
```diff
@@ -142,3 +149,315 @@ async fn test_light_client_proving() -> Result<()> {
...
+
+        let socket_addr = SocketAddr::new(rpc_bind_host.parse().unwrap(), rpc_bind_port);
+
+        let seq_test_client = make_test_client(socket_addr).await?;
+
+        let n_commitments = 4;
+
+        // publish min_soft_confirmations_per_commitment confirmations
+        for _ in 0..n_commitments * min_soft_confirmations_per_commitment {
+            // sequencer.client.http_client().eth_send_raw_transaction(data)
```
Comment (yaziciahmet) [inline bin/citrea/tests/bitcoin_e2e/light_client_test.rs]: remove comment

File: bin/citrea/tests/bitcoin_e2e/light_client_test.rs [hunk 12]
```diff
@@ -142,3 +149,315 @@ async fn test_light_client_proving() -> Result<()> {
...
+        let commitments = batch_prover
+            .client
+            .http_client()
+            .get_sequencer_commitments_on_slot_by_number(commitment_l1_height)
+            .await
+            .unwrap()
+            .unwrap();
+        assert_eq!(commitments.len(), n_commitments as usize);
+
+        // Ensure that batch proofs is submitted to DA (2x reveal & 2x commit txs)
```
Comment (yaziciahmet) [inline bin/citrea/tests/bitcoin_e2e/light_client_test.rs]: batch proofs are*

File: bin/citrea/tests/bitcoin_e2e/light_client_test.rs [hunk 13]
```diff
@@ -142,3 +149,315 @@ async fn test_light_client_proving() -> Result<()> {
...
+
+        // Since there are no batch proofs the state root should be the same as the last one
+        let light_client_proof2 = lcp2.unwrap();
+        assert_eq!(
+            light_client_proof2
+                .light_client_proof_output
+                .state_root
+                .to_vec(),
+            batch_proofs[1].proof_output.final_state_root
+        );
```
Comment (yaziciahmet) [inline bin/citrea/tests/bitcoin_e2e/light_client_test.rs]: irrelevant

**Review by yaziciahmet — CHANGES_REQUESTED** [review]

File: bin/citrea/tests/bitcoin_e2e/light_client_test.rs [hunk 14]
```diff
@@ -142,3 +149,307 @@ async fn test_light_client_proving() -> Result<()> {
...
+            .client
+            .http_client()
+            .get_sequencer_commitments_on_slot_by_number(commitment_l1_height)
+            .await
+            .unwrap()
+            .unwrap();
+        assert_eq!(commitments.len(), n_commitments as usize);
+
+        // Ensure that batch proofs are submitted to DA (2x reveal & 2x commit txs)
+        da.wait_mempool_len(4, Some(TEN_MINS)).await?;
```
Comment (yaziciahmet) [inline bin/citrea/tests/bitcoin_e2e/light_client_test.rs]: how are we sure that 4 sequencer commitments in the same DA block is going to generate 2 batch proofs?

File: bin/citrea/tests/bitcoin_e2e/light_client_test.rs [hunk 15]
```diff
@@ -142,3 +149,307 @@ async fn test_light_client_proving() -> Result<()> {
...
+
+        // Expect light client prover to have generated light client proof
+        let lcp = light_client_prover
+            .client
+            .http_client()
+            .get_light_client_proof_by_l1_height(batch_proof_l1_height)
+            .await?;
+        assert!(lcp.is_some());
+
+        let light_client_proof = lcp.unwrap();
```
Comment (yaziciahmet) [inline bin/citrea/tests/bitcoin_e2e/light_client_test.rs]: can just unwrap after await

File: bin/citrea/tests/bitcoin_e2e/light_client_test.rs [hunk 16]
```diff
@@ -142,3 +149,307 @@ async fn test_light_client_proving() -> Result<()> {
...
+        // Expect light client prover to have generated light client proof
+        let lcp2 = light_client_prover
+            .client
+            .http_client()
+            .get_light_client_proof_by_l1_height(last_finalized_height)
+            .await?;
+        assert!(lcp2.is_some());
+
+        // Since there are no batch proofs the state root should be the same as the last one
+        let light_client_proof2 = lcp2.unwrap();
```
Comment (yaziciahmet) [inline bin/citrea/tests/bitcoin_e2e/light_client_test.rs]: same

File: bin/citrea/tests/bitcoin_e2e/light_client_test.rs [hunk 17]
```diff
@@ -142,3 +149,307 @@ async fn test_light_client_proving() -> Result<()> {
...
+
+        // Expect light client prover to have generated light client proof
+        let lcp3 = light_client_prover
+            .client
+            .http_client()
+            .get_light_client_proof_by_l1_height(batch_proof_l1_height)
+            .await?;
+        assert!(lcp3.is_some());
+
+        let light_client_proof3 = lcp3.unwrap();
```
Comment (yaziciahmet) [inline bin/citrea/tests/bitcoin_e2e/light_client_test.rs]: same

**Review by yaziciahmet — DISMISSED** [review]

**Review by eyusufatik — APPROVED** [review]

**Pull request merged**